### PR TITLE
fix losing content while retrying

### PR DESCRIPTION
**Pull request by lly5044**
#2353

### CODE REVIEW

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/apache/RetryableRibbonLoadBalancingHttpClient.java [hunk 1]
```diff
@@ -96,7 +101,7 @@ public RibbonApacheHttpResponse execute(final RibbonApacheHttpRequest request, f
 
 		final RequestConfig requestConfig = builder.build();
 		final LoadBalancedRetryPolicy retryPolicy = loadBalancedRetryPolicyFactory.create(this.getClientName(), this);
-		RetryCallback retryCallback = new RetryCallback() {
+		RetryCallback<RibbonApacheHttpResponse, Exception> retryCallback = new RetryCallback<RibbonApacheHttpResponse, Exception>() {
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/apache/RetryableRibbonLoadBalancingHttpClient.java]: missing a space here

**Comment by lly5044** [comment]
I have to new a response object in RecoveryCallback and they have different constructors. I could not doing this without using reflection. 
Or I could set the response object into RetryableStatusCodeException instead of the raw reponse I set  now.
@ryanjbaxter

**Comment by spencergibb** [comment]
I'm a bit concerned with the supporting the complexity of this.

**Comment by ryanjbaxter** [comment]
@lly5044 here is what I was thinking regarding extracting the `RecoveryCallback` into its own class
```
public abstract class MyCallback<T, R> implements RecoveryCallback<T> {

	protected abstract T createResponse(R response, URI uri);

	@Override
	public T recover(RetryContext context) throws Exception {
		Throwable lastThrowable = context.getLastThrowable();
		if (lastThrowable != null && lastThrowable instanceof RetryableStatusCodeException) {
			RetryableStatusCodeException ex = (RetryableStatusCodeException) lastThrowable;
			return createResponse((R) ex.getResponse(), ex.getUri());
		}
		throw new RetryException("Could not recover", lastThrowable);
	}
}
```

Hope that makes things a bit clearer.

To @spencergibb's comment about complexity.  That was part of what I was trying to do by suggesting this is in hopes of making the code less complex and cleaner.

**Comment by lly5044** [comment]
I see, Thanks for your advice @ryanjbaxter

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonRecoveryCallback.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package org.springframework.cloud.netflix.ribbon;
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonRecoveryCallback.java]: Need copyright header

**Comment by lly5044** [comment]
any update on your discussion? @ryanjbaxter

**Comment by ryanjbaxter** [comment]
@lly5044 sorry for the delay.  I have actually been thinking about this change some more and how we can simplify not only this change but the whole retry functionality in Spring Cloud Netflix.  What I think I am going to do is incorporate your changes into what I want to do to simplify things, so I am most likely going to close this PR.  I hope to open a PR today or tomorrow with my changes and be sure to CC you on the changes so you can take a look.

**Comment by ryanjbaxter** [comment]
Closing in favor of #2685

**Pull request closed without merging**